Allow datasets to be used in taskflow

bolkedebruin opened the pull request:
Datasets could not be passed as parameters to taskflow functions as they could not be serialized. This commit:

1) changes the xcom serializer so that it now can serialize objects that have attr, dataclass or custom serializer
2) removes the need to for a custom serializer in lineage 
3) adds a version check to the serializer/deserializer 
4) registers any datasets as either inlets or outlets in the task
5) inlets/outlets that fail to serialize now raise an error instead of being silently ignored

@ashb @jedcunningham @kaxil 

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

uranusjr commented on airflow/decorators/base.py:
```diff
@@ -230,7 +230,7 @@ def _handle_output(self, return_value: Any, context: Context, xcom_push: Callabl
             self.outlets.append(return_value)
         if isinstance(return_value, list):
-            [self.outlets.append(item) for item in return_value if isinstance(item, Dataset)]
+            [self.outlets.append(item) for item in return_value if isinstance(item, Dataset)]  # type: ignore[func-returns-value]
```
Let’s use a normal for-loop for this. Using a list comprehension in this context is not common in the code base and may cause misunderstandings.

bolkedebruin replied:
I think the list comprehension is more clear than a normal for loop and the for loop is needlessly complicated.

bolkedebruin replied:
Okay I agree that it is not consistent with the rest of the code in this file.

ashb commented on airflow/decorators/base.py:
```diff
@@ -210,17 +212,27 @@ def __init__(
         super().__init__(task_id=task_id, **kwargs_to_upstream, **kwargs)
 
     def execute(self, context: Context):
+        for arg in chain(self.op_args, self.op_kwargs.values()):
+            if isinstance(arg, Dataset):
+                self.inlets.append(arg)
```
Shouldn't this be done in the prepare_lineage function (which I think will already called by `super().execute(context)` a few lines down.

bolkedebruin replied:
Are you suggesting that we would make this more generic? It now only affects the `DecoratedOperator` moving it to prepare_lineage (called from `pre_execute`) would make it affect all.

ashb replied:
That was my thought (specifically why should taskflow be treated specially?) but I haven't thought that deeply to consider what the implications might be.

bolkedebruin replied:
I don't think taskflow should be treated separately. It was my starting point (i.e. let taskflow support datasets). I can have a look at what it takes to move it a level higher.

ashb reviewed (commented):
Need to think about how this might affect loading from xcom existing values? Have you already considered that?

ashb commented on airflow/utils/json.py:
```diff
@@ -123,3 +136,89 @@ def dumps(self, obj, **kwargs):
+            return {
+                CLASSNAME: classname,
+                VERSION: version,
+                DATA: getattr(o.__class__, "serialize")(o)
```
Shouldn't this be

```suggestion
                DATA: getattr(o, "serialize")()
```

bolkedebruin replied:
That's not what stackoverflow is saying https://stackoverflow.com/a/38764817 ... but maybe you are right for `serialize` as it is an instance method. `deserialize` is a staticmethod

bolkedebruin replied:
Verified: it doesn't work for just `o` - tests do not pass.

ashb commented on airflow/utils/json.py:
```diff
@@ -40,6 +43,16 @@
 
 log = logging.getLogger(__name__)
 
+CLASSNAME = "__classname__"
+VERSION = "__version__"
+DATA = "__data__"
```
Rather than creating yet another way of encoding this kind of thing in the code base, is it possible we could reuse the `__var` and `__type` keys/prefixes that the BaseSerialization uses, or would that cause problems?

bolkedebruin replied:
as mentioned `__type` is overloaded and as the deserializer receives the nested dict first it is hard to distinguish between actual type and encoding type, which is a builtin.  Futhermore, I do think __var and __type are misnomers `__type` is not pythonic and `__var` is incorrect imho (it is the data of an object, not the a variable).

ashb commented on airflow/utils/json.py:
```diff
@@ -123,3 +136,89 @@ def dumps(self, obj, **kwargs):
+
+            cls = import_string(dct[CLASSNAME])
+
+            version = getattr(cls, "version", 0)
```
```suggestion
            version = getattr(cls, "version", DEFAULT_VERSION)
```

ashb commented on airflow/utils/json.py:
```diff
@@ -123,3 +136,89 @@ def dumps(self, obj, **kwargs):
+                )
+
+            if hasattr(cls, "deserialize"):
+                return getattr(cls, "deserialize")(dct[DATA])
```
We should probably pass version here too?

bolkedebruin replied:
yes

bolkedebruin replied:
Done. Version checks improved and tested.

ashb commented on tests/lineage/test_lineage.py:
```diff
@@ -122,10 +130,7 @@ def test_lineage_render(self, dag_maker):
         assert op1.inlets[0].url == f1s.format(DEFAULT_DATE)
         assert op1.outlets[0].url == f1s.format(DEFAULT_DATE)
 
-    def test_non_attr_outlet(self, dag_maker):
-        class A:
-            pass
```
This test was important to make sure that non-attr managed classes in Inlets/outlets continue to work.

bolkedebruin replied:
Non attr (or actually non serializable) inlets/outlets didn't work, they were just ignored. I don't think that is correct behavioral integrity. This leads to surprises where you supplied something as an outlet somewhere and made it an inlet to something else and it never 'arrived' an you don't know why.

ashb commented on tests/utils/test_json.py:
```diff
@@ -25,9 +25,26 @@
+
+    @staticmethod
+    def deserialize(data: dict):
+        return Z(data["x"])
```
I wonder if we should call these `airflow_serialize` etc, to reduce the risk of these clashing when used with "existing" objects?

bolkedebruin replied:
My idea is that serialization isn't something special and not necessarily tied to Airflow. In essence it is just a to_dict()/from_dict() which goes for anything in python.

ashb reviewed (commented):
Looks good so far, but I've just noticed/realised that this would mean that `orm_deserialize_value` (which currently just calls `deserialize_value`) would import user/dag-author specified code! We'll have to change the impl of orm_deserialize_value so we don't do that.

ashb commented on airflow/decorators/base.py:
```diff
@@ -207,17 +209,29 @@ def __init__(
     def _handle_output(self, return_value: Any, context: Context, xcom_push: Callable):
         """
         Handles logic for whether a decorator needs to push a single return value or multiple return values.
+        It sets outlets if any datasets are found in the returned value(s)
```
Doc style nit. The first paragraph appears in the summary so there should be a blank line. Doesn't matter here as this is a private method so wouldn't appear in docs anyway, but a good habit to get in to.
```suggestion

        It sets outlets if any datasets are found in the returned value(s)
```

ashb commented on airflow/models/xcom.py:
```diff
@@ -620,32 +621,41 @@ def serialize_value(
+        except (ValueError, TypeError) as ex:
             log.error(
-                "Could not serialize the XCom value into JSON."
+                f"{ex}."
```
```suggestion
                "%s."
```

bolkedebruin replied:
That seems to go against the current style?

ashb replied:
I'm surprised this wasn't triggering lint issues -- the rule we use is to never use f-string in log messages.

ashb replied:
```
      - id: check-lazy-logging
        name: Check that all logging methods are lazy
        entry: ./scripts/ci/pre_commit/pre_commit_check_lazy_logging.py
```

bolkedebruin replied:
Okido :-P geez :-)

ashb commented on airflow/models/xcom.py:
```diff
@@ -620,32 +621,41 @@ def serialize_value(
                 " then you need to enable pickle support for XCom"
-                " in your airflow config."
+                " in your airflow config or make sure to decorate your"
+                " object with attr."
```
```suggestion
                " object with attr.",
                ex
```

ashb commented on airflow/utils/json.py:
```diff
@@ -40,6 +45,16 @@
+OLD_SOURCE = "__source"
+OLD_DATA = "__var"
+
+DEFAULT_VERSION = "0"
```
```suggestion
DEFAULT_VERSION = 0
```

I think

ashb commented on airflow/utils/json.py:
```diff
@@ -123,3 +138,113 @@ def dumps(self, obj, **kwargs):
+            if hasattr(cls, "deserialize"):
+                return getattr(cls, "deserialize")(dct[DATA], version)
+
+            if VERSION in dct and int(dct[VERSION]) > version:
```
We've already checked version is in the dict on L>194, and shouldn't version always be an int? (I guess being "safe" about the type isn't a bad thing though)
```suggestion
            if dct[VERSION] > version:
```

bolkedebruin replied:
we are getting this from XCom which has no type safety guarantees. So i'd like to fail here explicitely on the conversion and have a meaningful error (instead of str vs int)

ashb commented on airflow/utils/json.py:
```diff
@@ -123,3 +138,113 @@ def dumps(self, obj, **kwargs):
+
+            version = getattr(cls, "version", 0)
+            if hasattr(cls, "deserialize"):
+                return getattr(cls, "deserialize")(dct[DATA], version)
```
This is the wrong version -- we want to pass the version of the "row", not the current latest version
```suggestion
                return getattr(cls, "deserialize")(dct[DATA], dct[VERSION])
```

(But we probably need to do some re-factoring/re-ordering rather than just apply this change)

bolkedebruin replied:
Good catch.

ashb reviewed (commented):
LGTM, a few small points then we can merge this

BasPH commented on docs/apache-airflow/concepts/taskflow.rst:
```diff
@@ -81,6 +81,49 @@ To use logging from your task functions, simply import and use Python's logging
+to tasks need to be able to be serialized. Airflow out of the box supports all built-in types (like int or str) and it
+supports objects that are decorated with ``@dataclass`` or ``@attr.define``, but it will need some help if you want to
+pass custom objects or if you want to control serialization yourself. To do so add the ``serialize()`` method to your
+class and the staticmethod ``deserialize(data: dict, version: int)`` to your class. Like so:
```
```suggestion
TaskFlow uses XCom to pass variables to each task. This requires variables that are used as arguments
to tasks to be able to be serialized. Airflow out of the box supports all built-in types (like int or str) and it
supports objects that are decorated with ``@dataclass`` or ``@attr.define``, but it will need some help if you want to pass custom objects, or if you want to control serialization yourself. To do so, implement the ``serialize()`` method and ``deserialize(data: dict, version: int)`` staticmethod in your class. Like so:
```

bolkedebruin replied:
I *think* you are reformatting, but don't know exactly. Please reapply if you don't mind.

BasPH replied:
Did some rewording for clarity

bolkedebruin replied:
If you can do that again that would be appreciated, I am getting word blind here :-P

BasPH commented on docs/apache-airflow/concepts/taskflow.rst:
```diff
@@ -81,6 +81,49 @@ To use logging from your task functions, simply import and use Python's logging
 
 Every logging line created this way will be recorded in the task log.
 
+Passing Arbitrary Objects As Parameters
```
```suggestion
Passing Arbitrary Objects As Arguments
```

BasPH commented on docs/apache-airflow/concepts/taskflow.rst:
```diff
@@ -81,6 +81,49 @@ To use logging from your task functions, simply import and use Python's logging
+        def deserialize(data: dict, version: int):
+            if version > 1:
+                raise TypeError(f"version > {MyCustom.version}")
+            return MyCustom(data["x"])
```
Could you demonstrate a taskflow task using the `MyCustom` class in the example?

bolkedebruin replied:
Are you open to "Tabbed" examples yet? :-)

BasPH replied:
Referring to https://github.com/apache/airflow/pull/26486 :-) ?

bolkedebruin replied:
Yes, but I am not ready for that yet :-). Example added.

BasPH replied:
I meant an example using this custom class. I'm not reading anything in relation to datasets. As a user, what do I do with `MyCustom`?

BasPH reviewed (commented):
Please add an example to the docs of how to use taskflow + datasets.

bolkedebruin committed:
Allow datasets to be used in taskflow

Datasets could not be passed as parameters to taskflow
functions as they could not be serialized. This commit:

1) changes the xcom serializer so that it now can serialize
objects that have attr, dataclass or a custom serializer
2) removes the need to for a custom serializer in lineage
3) adds a version check to the serializer/deserializer
4) registers any datasets as either inlets or outlets in the task
5) inlets or outlets that cannot be serialized now raise an error

BasPH commented on docs/apache-airflow/concepts/taskflow.rst:
```diff
@@ -81,6 +81,104 @@ To use logging from your task functions, simply import and use Python's logging
+    now = pendulum.now()
+
+
+    @dag(start_date=now, schedule="daily", catchup=False)
```
```suggestion
    @dag(start_date=now, schedule="@daily", catchup=False)
```

bolkedebruin committed:
Update docs/apache-airflow/concepts/taskflow.rst

Co-authored-by: Bas Harenslak <BasPH@users.noreply.github.com>

ashb commented on docs/apache-airflow/concepts/taskflow.rst:
```diff
@@ -1,19 +1,19 @@
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
```
```suggestion
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at
```

ashb commented on docs/apache-airflow/concepts/taskflow.rst:
```diff
@@ -1,19 +1,19 @@
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
```
```suggestion
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.
```

ashb committed:
Apply suggestions from code review

ashb reviewed (approved):
LGTM
